### Multiple Linear regression 

#### Importing the Libearies

In [4]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt

#### Importing the Dataset

In [5]:
dataset = pd.read_csv('F://Data_Set/50_Startups.csv')

In [6]:
dataset.head()

R_DSpend  Administration  MarketingSpend       State     Profit
0  165349.20       136897.80       471784.10    New_York  192261.83
1  162597.70       151377.59       443898.53  California  191792.06
2  153441.51       101145.55       407934.54     Florida  191050.39
3  144372.41       118671.85       383199.62    New_York  182901.99
4  142107.34        91391.77       366168.42     Florida  166187.94

In [19]:
dataset.shape

(50, 5)

In [29]:
print('Total No of Rows in the Dataset = ',len(dataset)) 

Total No of Rows in the Dataset =  50


In [26]:
print('Total No of Columns in the Dataset = ',len(dataset.columns))

Total No of Columns in the Dataset =  5


In [20]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   R_DSpend        50 non-null     float64
 1   Administration  50 non-null     float64
 2   MarketingSpend  50 non-null     float64
 3   State           50 non-null     object 
 4   Profit          50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


Check any Null value present in the dataset or not

In [7]:
dataset.isnull().sum()

R_DSpend          0
Administration    0
MarketingSpend    0
State             0
Profit            0
dtype: int64

Check any Duplicate Row Present in the Dataset or Not

In [11]:
dataset.duplicated().sum()

0

Columns of the Dataset

In [12]:
dataset.columns

Index(['R_DSpend', 'Administration', 'MarketingSpend', 'State', 'Profit'], dtype='object')

Check any Objective Column Present in the Daraset or Not 

In [14]:
#for i in dataset:
    #if i == dataset.columns['objective']:
        #print(dataset.columns)

In [16]:
dataset.dtypes=='object'

R_DSpend          False
Administration    False
MarketingSpend    False
State              True
Profit            False
dtype: bool

In [32]:
dataset.corr()

R_DSpend  Administration  MarketingSpend    Profit
R_DSpend        1.000000        0.241955        0.724248  0.972900
Administration  0.241955        1.000000       -0.032154  0.200717
MarketingSpend  0.724248       -0.032154        1.000000  0.747766
Profit          0.972900        0.200717        0.747766  1.000000

#### Spliting the Dataset into Dependent and Independent Variable

Independent Variable

In [33]:
X = dataset.iloc[:,:-1].values

In [35]:
y = dataset.iloc[:,-1].values

#### Encoding the Categorical Data

In [37]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

In [38]:
ct = ColumnTransformer(transformers=[('encoder',OneHotEncoder(),[3])],remainder='passthrough')

In [39]:
X = np.array(ct.fit_transform(X))

In [40]:
X

array([[0.0, 0.0, 1.0, 165349.2, 136897.8, 471784.1],
       [1.0, 0.0, 0.0, 162597.7, 151377.59, 443898.53],
       [0.0, 1.0, 0.0, 153441.51, 101145.55, 407934.54],
       [0.0, 0.0, 1.0, 144372.41, 118671.85, 383199.62],
       [0.0, 1.0, 0.0, 142107.34, 91391.77, 366168.42],
       [0.0, 0.0, 1.0, 131876.9, 99814.71, 362861.36],
       [1.0, 0.0, 0.0, 134615.46, 147198.87, 127716.82],
       [0.0, 1.0, 0.0, 130298.13, 145530.06, 323876.68],
       [0.0, 0.0, 1.0, 120542.52, 148718.95, 311613.29],
       [1.0, 0.0, 0.0, 123334.88, 108679.17, 304981.62],
       [0.0, 1.0, 0.0, 101913.08, 110594.11, 229160.95],
       [1.0, 0.0, 0.0, 100671.96, 91790.61, 249744.55],
       [0.0, 1.0, 0.0, 93863.75, 127320.38, 249839.44],
       [1.0, 0.0, 0.0, 91992.39, 135495.07, 252664.93],
       [0.0, 1.0, 0.0, 119943.24, 156547.42, 256512.92],
       [0.0, 0.0, 1.0, 114523.61, 122616.84, 261776.23],
       [1.0, 0.0, 0.0, 78013.11, 121597.55, 264346.06],
       [0.0, 0.0, 1.0, 94657.16, 145077.58

Dependent Variable

In [44]:
Y = dataset.iloc[:,-1].values 

In [45]:
Y

array([192261.83, 191792.06, 191050.39, 182901.99, 166187.94, 156991.12,
       156122.51, 155752.6 , 152211.77, 149759.96, 146121.95, 144259.4 ,
       141585.52, 134307.35, 132602.65, 129917.04, 126992.93, 125370.37,
       124266.9 , 122776.86, 118474.03, 111313.02, 110352.25, 108733.99,
       108552.04, 107404.34, 105733.54, 105008.31, 103282.38, 101004.64,
        99937.59,  97483.56,  97427.84,  96778.92,  96712.8 ,  96479.51,
        90708.19,  89949.14,  81229.06,  81005.76,  78239.91,  77798.83,
        71498.49,  69758.98,  65200.33,  64926.08,  49490.75,  42559.73,
        35673.41,  14681.4 ])

#### Spliting the dataset into training set and test set

In [42]:
from sklearn.model_selection import train_test_split 

In [46]:
x_train,x_test,y_train,y_test = train_test_split(X,Y,train_size=0.8,random_state=0)

In [47]:
x_train

array([[0.0, 1.0, 0.0, 55493.95, 103057.49, 214634.81],
       [0.0, 0.0, 1.0, 46014.02, 85047.44, 205517.64],
       [0.0, 1.0, 0.0, 75328.87, 144135.98, 134050.07],
       [1.0, 0.0, 0.0, 46426.07, 157693.92, 210797.67],
       [0.0, 1.0, 0.0, 91749.16, 114175.79, 294919.57],
       [0.0, 1.0, 0.0, 130298.13, 145530.06, 323876.68],
       [0.0, 1.0, 0.0, 119943.24, 156547.42, 256512.92],
       [0.0, 0.0, 1.0, 1000.23, 124153.04, 1903.93],
       [0.0, 0.0, 1.0, 542.05, 51743.15, 0.0],
       [0.0, 0.0, 1.0, 65605.48, 153032.06, 107138.38],
       [0.0, 0.0, 1.0, 114523.61, 122616.84, 261776.23],
       [0.0, 1.0, 0.0, 61994.48, 115641.28, 91131.24],
       [1.0, 0.0, 0.0, 63408.86, 129219.61, 46085.25],
       [1.0, 0.0, 0.0, 78013.11, 121597.55, 264346.06],
       [1.0, 0.0, 0.0, 23640.93, 96189.63, 148001.11],
       [1.0, 0.0, 0.0, 76253.86, 113867.3, 298664.47],
       [0.0, 0.0, 1.0, 15505.73, 127382.3, 35534.17],
       [0.0, 0.0, 1.0, 120542.52, 148718.95, 311613.29],
       

In [48]:
x_test

array([[0.0, 1.0, 0.0, 66051.52, 182645.56, 118148.2],
       [1.0, 0.0, 0.0, 100671.96, 91790.61, 249744.55],
       [0.0, 1.0, 0.0, 101913.08, 110594.11, 229160.95],
       [0.0, 1.0, 0.0, 27892.92, 84710.77, 164470.71],
       [0.0, 1.0, 0.0, 153441.51, 101145.55, 407934.54],
       [0.0, 0.0, 1.0, 72107.6, 127864.55, 353183.81],
       [0.0, 0.0, 1.0, 20229.59, 65947.93, 185265.1],
       [0.0, 0.0, 1.0, 61136.38, 152701.92, 88218.23],
       [0.0, 1.0, 0.0, 73994.56, 122782.75, 303319.26],
       [0.0, 1.0, 0.0, 142107.34, 91391.77, 366168.42]], dtype=object)

In [49]:
y_train

array([ 96778.92,  96479.51, 105733.54,  96712.8 , 124266.9 , 155752.6 ,
       132602.65,  64926.08,  35673.41, 101004.64, 129917.04,  99937.59,
        97427.84, 126992.93,  71498.49, 118474.03,  69758.98, 152211.77,
       134307.35, 107404.34, 156991.12, 125370.37,  78239.91,  14681.4 ,
       191792.06, 141585.52,  89949.14, 108552.04, 156122.51, 108733.99,
        90708.19, 111313.02, 122776.86, 149759.96,  81005.76,  49490.75,
       182901.99, 192261.83,  42559.73,  65200.33])

In [50]:
y_test

array([103282.38, 144259.4 , 146121.95,  77798.83, 191050.39, 105008.31,
        81229.06,  97483.56, 110352.25, 166187.94])

#### Training the Model

In [51]:
from sklearn.linear_model import LinearRegression

In [52]:
lr = LinearRegression()

Training the Model

In [53]:
lr.fit(x_train,y_train)

LinearRegression()

#### Predict the Test set result

In [54]:
y_pred = lr.predict(x_test)

In [55]:
y_pred

array([103015.20159796, 132582.27760815, 132447.73845173,  71976.09851258,
       178537.48221051, 116161.24230163,  67851.69209676,  98791.73374689,
       113969.43533011, 167921.06569547])

In [59]:
df = pd.DataFrame(y_test,y_pred)

In [57]:
df

0
103015.201598  103282.38
132582.277608  144259.40
132447.738452  146121.95
71976.098513    77798.83
178537.482211  191050.39
116161.242302  105008.31
67851.692097    81229.06
98791.733747    97483.56
113969.435330  110352.25
167921.065695  166187.94

Making a single prediction (for example the profit of a startup with R&D Spend = 160000, Administration Spend = 130000, Marketing Spend = 300000 and State = 'California')

In [61]:
lr.predict([[1, 0, 0,160000,130000,300000]]) 

array([181566.92389383])

In [63]:
print(lr.coef_)

[ 8.66383692e+01 -8.72645791e+02  7.86007422e+02  7.73467193e-01
  3.28845975e-02  3.66100259e-02]


In [64]:
print(lr.intercept_)

42467.529248581035


Therefore, the equation of our multiple linear regression model is:

**Profit=86.6×Dummy State 1−873×Dummy State 2+786×Dummy State 3+0.773×R&D Spend+0.0329×Administration+0.0366×Marketing Spend+42467.53**